linkml/generators/pydanticgen/array.py

```diff
@@ -0,0 +1,396 @@
+            raise ValueError("Can only specify EITHER exact_cardinality OR minimum/maximum cardinality")
+        elif dimension.exact_cardinality:
+            dmin = dimension.exact_cardinality
+            dmax = dimension.exact_cardinality
+        elif dimension.minimum_cardinality or dimension.maximum_cardinality:
+            dmin = dimension.minimum_cardinality
+            dmax = dimension.maximum_cardinality
+        else:
+            # TODO: handle labels for labeled but unshaped arrays
+            return SlotResult(annotation="List[" + dtype + "]")
+
+        items = []
+        if int(PYDANTIC_VERSION[0]) >= 2:
+            if dmin is not None:
+                items.append(f"min_length={dmin}")
+            if dmax is not None:
+                items.append(f"max_length={dmax}")
+        else:
+            if dmin is not None:
+                items.append(f"min_items={dmin}")
+            if dmax is not None:
+                items.append(f"max_items={dmax}")
+        items.append(f"item_type={dtype}")
+        items = ", ".join(items)
+        annotation = f"conlist({items})"
+
+        return SlotResult(annotation=annotation, imports=_ConListImports)
+
+    def any_shape(self, array: Optional[ArrayExpression] = None, with_inner_union: bool = False) -> SlotResult:
+        """
+        An AnyShaped array
+
+        Args:
+            array (:class:`.ArrayExpression`): The array expression (not used)
+            with_inner_union (bool): If ``True`` , the innermost type is a ``Union`` of the ``AnyShapeArray`` class
+                and ``dtype`` (default: ``False`` )
+
+        """
+        if self.dtype in ("Any", "AnyType"):
+            annotation = "AnyShapeArray"
+        else:
+            annotation = f"AnyShapeArray[{self.dtype}]"
+
+        if with_inner_union:
+            annotation = f"Union[{annotation}, {self.dtype}]"
+        return SlotResult(annotation=annotation, injected_classes=_AnyShapeArrayInjects, imports=_AnyShapeArrayImports)
+
+    def anonymous_shape(self, array: ArrayExpression) -> SlotResult:
+        if array.exact_number_dimensions:
+            return SlotResult(annotation=self._list_of_lists(array.exact_number_dimensions, self.dtype))
+        elif not array.maximum_number_dimensions and (
+            array.minimum_number_dimensions is None or array.minimum_number_dimensions == 1
+        ):
+            return self.any_shape()
+        elif array.maximum_number_dimensions:
+            min_dims = array.minimum_number_dimensions if array.minimum_number_dimensions is not None else 1
+            annotations = [
+                self._list_of_lists(i, self.dtype) for i in range(min_dims, array.maximum_number_dimensions + 1)
+            ]
+            # TODO: Format this nicely!
+            return SlotResult(annotation="Union[" + ", ".join(annotations) + "]")
+        else:
+            return SlotResult(
```

Comment:
```suggestion
            # min specified with no max
            # e.g., if min = 3, annotation = List[List[AnyShapeArray[dtype]]]
            return SlotResult(
```

Just adding comments and examples to help me and future code readers

Reply:
Adding examples at the risk of the example becoming outdated and inconsistent, but I think this one will not...

Reply:
i like comments :) and i think that a lot of these will be made clearer when i finish adding the docs, which will be a lot of "this syntax generates that type annotation" examples.